src/SignalR/server/Core/src/Internal/HubMethodDescriptor.cs

```diff
@@ -27,10 +34,28 @@ public HubMethodDescriptor(ObjectMethodExecutor methodExecutor, IEnumerable<IAut
                 ? MethodExecutor.AsyncResultType
                 : MethodExecutor.MethodReturnType;
 
-            if (IsChannelType(NonAsyncReturnType, out var channelItemType))
+            foreach (var returnType in NonAsyncReturnType.AllBaseTypes().Concat(NonAsyncReturnType.GetInterfaces()))
             {
-                IsChannel = true;
-                StreamReturnType = channelItemType;
+                if (!returnType.IsGenericType)
+                {
+                    continue;
+                }
+
+                var openReturnType = returnType.GetGenericTypeDefinition();
+
+                if (openReturnType == typeof(IAsyncEnumerable<>))
```

Comment:
If a type inherits from ChannelReader and has IAsyncEnumerable, is there a preference between the two? Currently base types will come before the interfaces so channel reader is preferred.

Reply:
I switched the preference to IAsyncEnumerable and added a test.